### ISSUE TITLE
Implement loops with ipairs

### ISSUE THREAD

**Issue by hugomg** [pallene-lang/pallene#164]
It would be nice to be able to use ipairs in Pallene:

```
for i, x in ipairs(xs) do
    -- do something
end
```

Implementing general for-in loops as well would be a nice bonus. But the particular case of ipairs is what I miss the most.

Two things to watch out for when implementing this:

1. `ipairs` is not a keyword. It can be shadowed by local declarations.
2. The ipairs loop is not equivalent to `for i=1, #xs`, if there are holes in the array. The ipairs version always stops at the first hole, while the `#xs` version may or may not do that.

**Comment by Akylzhan** [comment]
I can work on this issue but I do not know how to do this. Can you guide me?

**Comment by ahmedkrmn** [comment]
Hi @Akylzhan.
What do you have in mind to implement this issue?
We can discuss it together to figure out a solution.

**Comment by Akylzhan** [comment]
@ahmedkrmn I did not look the source code in details and I am quite clueless, but the only things that comes to my mind is to create new command for ```gen_cmd``` in ```coder.lua```. Though as @hugomg mentioned, ```ipairs``` should not be keyword; therefore this is not good solution.

Sorry, if I am mistaken

**Comment by ahmedkrmn** [comment]
`ipairs` is currently not a reserved keyword in the language, you can
declare a variable that has the name of `ipairs`.
I believe the first step is to add `ipairs` to the list of reserved words
during the transpiler lexical analysis phase.

**Comment by Akylzhan** [comment]
as I understood, I need to first declare 'ipairs' function in ```pallene/builtins.lua``` and then declare it in ```pallene/ir.lua``` in table ```ir_cmd_constructors``` as ```BuiltinIpairs``` (but did not get what should I include to this table element). Then, add new ```elseif``` after line 419 of ```pallene/to_ir.lua```. Right?

**Comment by ahmedkrmn** [comment]
@Akylzhan before jumping into implementation details, you first have to add `ipairs` to the list of reserved keywords found in `/pallene/lexer.lua` then you should modify the language grammar in `/pallene/parser.lua` to allow for the new syntax introduced by having `ipairs` in for loops.

**Comment by Akylzhan** [comment]
@ahmedkrmn I thought that I need to add ipairs as built_in function

**Comment by hugomg** [comment]
Hi, @Akylzhan, I'm glad to see that you are interested in getting involved in Pallene.  To solve this particular issue, I imagine that it could be broken down in two main tasks:

First one would be to add support for `for in` loops:

1. Extend the parser with so it can parse `for in` loops
2. Extend the IR so it can represent `for in` loops
3. Extend the code generation (coder.lua) to generate code for the `for in` loops

At this point, there is no ipairs yet, but it would already be possible to test the `for in` feature with our own iterator functions. Once that is done, we could move on to the second task, which would be to add support for the ipairs function:

1. Add ipairs as a built-in function
2. Modify the compiler to recognize the common case of `for i, x in ipairs(xs)`, and optimize it accordingly.

Going back to the question, @Akylzhan, if you want to play around a good place to start would be try to download and compile Pallene in you own machine, and get to the point where you can run the test suite. Then we could talk more about what would need to be done to extend the compiler with `for in` loops, which would be the first step. Let me know if you have any issues with that. We're always working on improving the documentation, so any common stumbling blocks would be good for us to know.

**Comment by Akylzhan** [comment]
The problem is that I don't know how to show parser to accept for-loop with ```in``` OR accept traditional for-loop

**Comment by hugomg** [comment]
We're not planning on making "ipairs" into a reserved keyword, so there will be no changes on the lexer.

> as I understood, I need to first declare 'ipairs' function [...]

Once we add the ipairs function, that is the basic idea of how to do do it. But first we need to extend the language with `for in` loops...

> The problem is that I don't know how to show parser to accept for-loop with

We'll need to add a new case for the for-in loops in the ast.lua, and extend the parser in parser.lua. I can help you with that. Are you already familiar with LPeg?

**Comment by Akylzhan** [comment]
I am not familiar with LPeg. But as I understood I need to add this:
```For    = {"loc", "decl", "fields", "block"},``` in ast.lua under original For table.
And, then, add this:```
                     / (P  FOR decl^DeclFor
                           IN explist
                           DO^DoFor block END^EndFor)            -> StatFor
``` to parser.lua under original For

But I am doing smth wrong

**Comment by hugomg** [comment]
You are on the right track.

> I need to add this: For = {"loc", "decl", "fields", "block"}, in ast.lua under original For table.

The name of each different case needs to be unique, because it is used later when we pattern match against the `_tag`. In this case I'd rename the current `For` into `ForNum`, and name the new for loop `ForIn`.

It might be easier to rename the current `For` into `ForNum` in the first commit, making sure that all the busted tests still pass. Then we can move on to adding `ForIn`.

> explist

For now, let's pretend that `ipairs` returns a single value, that is let's pretend that it is `for x in ipairs(xs)` instead of `for i, x in ipairs(xs)`. Multiple return values depend on PR #177 getting merged first.

> under original For. But I am doing smth wrong

Working with the parser is subtle. We're using [lpeglabel](https://github.com/sqmedeiros/lpeglabel/), which is an extension of LPeg with support for error handling. If a token or nonterminal is followed by a `^`, then instead of backtracking and attempting to match something else, the parser will instead immediately abort with an error.

This means that the parser doesn't have a chance to try to parse the code as a for-in loop, because it errors out while trying to parse it as a numeric loop. 

The first step of the fix is that we have to remove the error handling from the numeric for loop parsing, so that the parser is allowed to backtrack if it finds a `for x` but it is followed by an `=` instead of an `in`.

```
 / (P  FOR decl ASSIGN
                      exp^Exp1For
                      COMMA^CommaFor exp^Exp2For
                      (COMMA exp^Exp3For)?                  -> opt
                      DO^DoFor block END^EndFor)            -> StatForNum
```

Then we can add the block of code for parsing the for-in loop after this.

``` 
 / (P  FOR decl IN
          exp^ExpForIn
          DO^DoFor block END^EndFor)            -> StatFor
```

And to finish it, we have to add a third case handles those error cases that we stopped taking care of in the first cases. That is, it should produce an error if there is a missing variable declaration after the `for`, or if there is not a `=` or an `in` after the variable declaration.

```
/ (FOR decl^DeclFor (ASSIGN / IN)^ForInOrAssign
```

The error messages associated with these errors are in the `syntax_errors.lua` file. While we are making these changes, it may be necessary to edit, add, or remove some of the syntax errors. We should also update the test cases accordingly.

**Comment by hugomg** [comment]
By the way, feel free to open a PR once you have the first commit working. That way we can follow along from the start.

**Comment by Akylzhan** [comment]
@hugomg Hello, can you help me? I have added the definitions for for-in loop in ast and wanted to change 'pallene/checker.lua' file but do not understand how to do it

**Comment by hugomg** [comment]
After thinking a bit more about the issue here, I think we might need to go back a bit and rethink our strategy. In the curreny version of Pallene's type system, modelling the `for-in` loop in its generality is a bit unwieldy. It might be easier if we for now restrict for-in loops to loops that are literal `for i, x in ipairs(xs)` loops.

The first change would be to modify the parser so that it accepts a comma-separated list of decls instead of a single decl, as we do now.

Then, we would have to introduce `ipairs` as a builtin function in builtins.lua. For now, we can declare it's type as being a function that receives a single argument of type `{any}` and returns no arguments. 

Once we get that working we can go back to the checkwe, where we are currently stuck.

--------------

> But have troubles with it, cannot convert 'ast.Exp.Integer' to 'types.T.Integer' in order to declare local for-loop variable

Would you mind giving a bit more context on how you arrived to this problem? That way I might be able to give a better answer.

What I can say for now is that `ast.Exp` is the datatype we use to represent nodes in our abstract syntax tree, and `ast.Exp.Integer` corresponds to integer literals such as `0`, `1`, or `17`. Meanwhile, `types.T` is the datatype we use to represent Pallene types. `Types.T.Integer` corresponds to the `integer` type. By the way, the `types.T` datatype should not be confused with `ast.Type`, which are the nodes in the syntax tree that represent that type annotations.

**Comment by Akylzhan** [comment]
Thank you very much for your responses, you helped me a lot. Now I have added the definition of 'ipairs'. But I have a question, if Pallene currently does not support function with multiple return, then, how are we supposed to implement it? Or, I am mistaken?

**Comment by hugomg** [comment]
One thing we can do for now is to recognize `for i, x in ipairs(xs)` as a special case, and forbid the use of `ipairs` anywhere else, with a compile-time error. It wouldn't be elegant, but would cover the most common use cases.

If we do this we wouldn't have to actually implement `ipairs` as a function. The idea is that we would be able to compile `for i, x in ipairs(xs)` into a simple loop which never calls any iterator functions.

**Comment by Akylzhan** [comment]
Hello, @hugomg! I have stuck with several problems, I could not come with the idea how to assign table value to second variable of loop, the only solution that came to my mind is using ```ir.Cmd.GetArr``` but it does not accept ```checker.Name.Local``` variable or id of variable. The next problem is similar, I don't know how increasing index variable by one, I though that I should use ir.Cmd.Binop but it does not accept id or checker.Name.Local variable. Can you help me again?

**Comment by hugomg** [comment]
Hi, could you provide a small code snippet of one of the things that you tried and didn't work?


Em 11 de março de 2020 06:56:37 BRT, Akylzhan Sauranbay <notifications@github.com> escreveu:

**Comment by Akylzhan** [comment]
I have updated my branch

**Comment by srijan-paul** [comment]
Hi there, I'd like to work on this issue after playing around with Pallene a bit more. It looks a lot of the prelims required to support has already been put in place after this issue was opened. 

One approach I can think of is to ipairs like as a `builtin` so it can be shadowed by local declarations, then we could bake in some code in this section of `parser.lua`: 

```lua
elseif self:peek("for") then 
-- // try('=')
 elseif self:peek(",") or self:peek("in") then
            local decls = {decl1}
     	    -- 
            if self.peek("NAME") and self.doublepeek("in") then
		-- mark the loop as an ipairs loop
		-- if no local overload is found.
	    end
	    --
            while self:try(",") do
                table.insert(decls, self:Decl())
            end
            local _    = self:e("in")
            local exps = self:ExpList1()
            local _    = self:e("do"); self:loop_begin()
            local body = self:Block(); self:loop_end()
            local _    = self:e("end", start)
            self:syntax_error(start.loc, "for-in loops are not implemented yet")
            return ast.Stat.ForIn(start.loc, decls, exps, body)

        else
```

Now I'll admit I'm a bit unfamiliar with Pallene's `coder.lua` and `ir.lua`, so I'll have to familiarize myself with that first to carry this over to codegen. If I could get a heads up on how one would  go about implementing this in the backend with Lua's C API with `PALLENE_FLT_FOR_LOOP_BEGIN`, I could start work on this.

**Comment by hugomg** [comment]
Hi srijan, glad to see you are interested on working on this.

> mark the loop as an ipairs loop if no local overload is found.

The parser isn't the best place to do this, because at that part of the program we are not tracking what names are in scope. That is the responsibility of the checker.lua, which adds a _name annotation to various AST nodes. This annotation can then be used by the checker and the to_ir passes.

One possible direction would be something like this:

1. Rename the `Stat.For` in the `ast.lua` to `ast.Stat.ForNum` and introduce a new `ast.Stat.ForIn`.  (like we did in the first attempt at solving this issue)
2. Change the parser.lua to emit these ast.Stat.ForIn nodes instead of raising a "for-in loops are not implemented yet" error. Update the test cases accordingly.

After this first steps we could already open a draft PR to start seeing the direction things are going. Moving on...

3. Add an `ipairs` function in the  builtins.lua. (This should be possible now that functions are allowed to return multiple values)
4. In the `ir.lua`, create a new kind Cmd for the ipairs loop. (And maybe also a new kind of command for generic for-in loops)
5. The to_ir step is responsible for converting the AST into the IR. When trasnlating an ast.Stat.ForIn we can look at the `_name` annotation to see if the argument to the for-in loop is a call to ipairs which hasn't been shadowed. If it is we generate the IR for an ipairs loop. Otherwise we generate the IR for a generic for loop. (It is also OK if we leave generic for loops unimplemented at first).

> If I could get a heads up on how one would go about implementing this in the backend with Lua's C API with PALLENE_FLT_FOR_LOOP_BEGIN, I could start work on this.

One thing that I like doing is start by writing the C code "by hand". You can use --emit-c to generate a C file, edit that by hand and then use --compile-c to generate an executable. Once you figure out a way to implement the thing you want to implement then you can go back to coder.lua and teach the C code generator how to generate that kind of code.

> PALLENE_FLT_FOR_LOOP_BEGIN

The idea behind the for-loop-begin macro is that numeric for loops have some special logic to handle integer overflow. We put that inside a C macro to make the generated C code a bit more readable.

Now that you mention it, we will probably want to rename it to FORNUM_LOOP, to make it clear that it doesn't have anything to do with the for-in loops.

**Comment by srijan-paul** [comment]
Yeah I did play around a bit with --emit-c. Though I have to get a bit familiar with it. 
I like your approach, sounds straight forward. Thanks for explaining.

So I guess now I should implement `ForNum` and `ForIn` AST nodes, submit a draft PR and we continue from there ? Sound good ?

**Comment by hugomg** [comment]
Yup, that sounds like a good place to start.

**Comment by srijan-paul** [comment]
Hi there, sorry about the late update, was caught up with exams.
So I renamed the `ast.Stat.For` to `ast.Stat.ForNum`, made the parser return `ForNum` nodes and added a test.
Should I change the checker accordingly or open a draft PR before making changes in other passes ?

**Comment by hugomg** [comment]
Oh, glad to hear back from you.

Feel free to create a draft PR for it. Getting early feedback can be very helpful.

**Comment by hugomg** [comment]
Closed by #346
